### DISCUSSION TITLE
Can qpdf do font subsetting?

### DISCUSSION THREAD

**Discussion by concreteRick** [qpdf/qpdf#1001] [Q&A]
There's nothing in the docs about this, but I'm asking anyway, just in case ...

There are two scenarios. The first, simpler case is a PDF that has the entire Arial font (for example)
embedded, but only uses eg. about 10% of the Arial charset. The second scenario is a PDF comprising
multiple other PDFs assembled together/merged, each of them having the entire Arial font embedded.

In scenario two, imagine a PDF that includes dozens of other PDF images originating from a conversion
or merging process. In this case the entire Arial font is included dozens of times, resulting
truly bloated PDF. The required subsetting in this situation is I guess very much more complicated,
and is applicable regardless of whether the fonts in the included PDFs are present in their entireity
or as subsets.

This option:

        --compress-streams=y

does a fantastic job, especially wrt tagged PDF. But font streams don't seem to be subject to compression
and this is a different challenge in any case.

The Optimize PDF option in Adobe Acrobat Pro generally does a good job, but even that app. sometimes
has difficulties in the case of large files with dosens of repeated instances of the same fonts.

If not already there, is this a feature that might be added to qpf in the future, or is it way off the
roadmap?

I could not find any comparable, similarly licensed, C++ or Rust apps or libraries advertising this kind
of capability.

**Comment by concreteRick** [comment]
Checking further, this might be related to:
https://github.com/qpdf/qpdf/issues/448
??

**Comment by concreteRick** [comment]
Sorry, closed and marked as answer by mistake.

**Comment by jberkenbilt** [comment]
qpdf does not do font subsetting. That would require qpdf to actually know how to work with fonts, something that could be done with a library like freetype. It's unlikely that qpdf will have this capability in the foreseeable future, if ever. However if someone wanted to write something that did font subsetting, it would certainly be possible to use qpdf and freetype to do it, but it would require low-level interpretation of the PDF.
